### PR TITLE
refactor auth interface to return error when verify failure

### LINKED ISSUE #20563: `docker login` reports sucessful login with incorrect password / token
### Description

```
$ docker login localhost:3000
Username: foobar
Password: 
Login Succeeded
```
↑ this should fail as i entered gibberish for the password

```
$ docker push localhost:3000/foobar/qgis
Using default tag: latest
The push refers to repository [localhost:3000/foobar/qgis]
5a1bd4bee150: Layer already exists 
1b04ddff91f2: Layer already exists 
4dd0c5812fd4: Layer already exists 
unauthorized: authentication required
```
↑ ACLs seem to generally work at least ;)

### Gitea Version

692707f14519b677de4601e5f40469989f82eed6

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

build from source

### Database

SQLite

### PR DESCRIPTION
**Pull request by lunny**
This PR changed the Auth interface signature from 
`Verify(http *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *user_model.User` 
to 
`Verify(http *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error)`. 

There is a new return argument `error` which means the verification condition matched but verify process failed, we should stop the auth process.

Before this PR, when return a `nil` user, we don't know the reason why it returned `nil`. If the match condition is not satisfied or it verified failure? For these two different results, we should have different handler. If the match condition is not satisfied, we should try next auth method and if there is no more auth method, it's an anonymous user. If the condition matched but verify failed, the auth process should be stop and return immediately.

This will fix #20563

### CODE REVIEW

File: modules/context/context.go [hunk 1]
```diff
@@ -662,7 +662,13 @@ func getCsrfOpts() CsrfOptions {
 // Auth converts auth.Auth as a middleware
 func Auth(authMethod auth.Method) func(*Context) {
 	return func(ctx *Context) {
-		ctx.Doer = authMethod.Verify(ctx.Req, ctx.Resp, ctx, ctx.Session)
+		var err error
+		ctx.Doer, err = authMethod.Verify(ctx.Req, ctx.Resp, ctx, ctx.Session)
+		if err != nil {
+			log.Error("Failed to verify user %v: %v", ctx.Req.RemoteAddr, err)
```
Comment (delvh) [inline modules/context/context.go:668]: Wait, we don't log API verification failures?
This line is missing in `APIAuth`…

File: services/packages/auth.go [hunk 2]
```diff
@@ -41,9 +41,14 @@ func CreateAuthorizationToken(u *user_model.User) (string, error) {
...
-	parts := strings.SplitN(req.Header.Get("Authorization"), " ", 2)
+	h := req.Header.Get("Authorization")
+	if h == "" {
+		return 0, nil
+	}
+
+	parts := strings.SplitN(h, " ", 2)
 	if len(parts) != 2 {
-		return 0, fmt.Errorf("no token")
+		return 0, fmt.Errorf("split token failed: %s", h)
```
Comment (zeripath) [inline services/packages/auth.go]: This may cause the leak of security data

Comment (lunny) [reply services/packages/auth.go]: done.

File: services/auth/sspi_windows.go [hunk 3]
```diff
@@ -77,15 +77,15 @@ func (s *SSPI) Free() error {
...
 	if !s.shouldAuthenticate(req) {
-		return nil
+		return nil, nil
 	}
 
 	cfg, err := s.getConfig()
 	if err != nil {
 		log.Error("could not get SSPI config: %v", err)
-		return nil
+		return nil, nil
```
Comment (KN4CK3R) [inline services/auth/sspi_windows.go]:
```suggestion
		return nil, err
```
There is an `err`.

**Comment by KN4CK3R** [comment]
For clarification: The Docker login problem is now fixed because the Basic Auth will return an error which prevents the execution of the Docker code?

**Comment by lunny** [comment]
> For clarification: The Docker login problem is now fixed because the Basic Auth will return an error which prevents the execution of the Docker code?

Yes, on previous code. wrong password will return nil so Gitea think it's an anonymous login, but now we have a different result for anonymous login and wrong password.

**Review by KN4CK3R — APPROVED** [review]

File: services/auth/sspi_windows.go [hunk 4]
```diff
@@ -108,32 +108,32 @@ func (s *SSPI) Verify(req *http.Request, w http.ResponseWriter, store DataStore,
 			log.Error("%v", err)
```
Comment (delvh) [inline services/auth/sspi_windows.go:108]: Erm… Given that we now pass this error down, should we still log it here?

File: services/auth/basic.go [hunk 5]
```diff
@@ -77,11 +77,11 @@ func (b *Basic) Verify(req *http.Request, w http.ResponseWriter, store DataStore
 		u, err := user_model.GetUserByID(req.Context(), uid)
 		if err != nil {
 			log.Error("GetUserByID:  %v", err)
```
Comment (delvh) [inline services/auth/basic.go:79]: Regarding all these logs:
I've seen quite often now that they are still there.
Shouldn't we pass all of them down to their callers?

File: services/auth/group.go [hunk 6]
```diff
@@ -80,23 +79,23 @@ func (b *Group) Free() error {
 }
 
 // Verify extracts and validates
-func (b *Group) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *user_model.User {
-	if !db.HasEngine {
-		return nil
-	}
```
Comment (delvh) [inline services/auth/group.go:86]: I have the slight suspicion that this check was there for a reason.
Maybe it is not set sometimes, and then our program will panic?

File: services/auth/group.go [hunk 7]
```diff
@@ -80,23 +79,23 @@ func (b *Group) Free() error {
...
-
+func (b *Group) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error) {
 	// Try to sign in with each of the enabled plugins
 	for _, ssoMethod := range b.methods {
-		user := ssoMethod.Verify(req, w, store, sess)
+		user, err := ssoMethod.Verify(req, w, store, sess)
+		if err != nil {
+			return user, err
```
Comment (wolfogre) [inline services/auth/group.go]:
```suggestion
			return nil, err
```

**Review by wolfogre — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Review by delvh — COMMENTED** [review]
I've also noticed a few things in an unreleased review before the PR was merged.
Here they are:

**Pull request merged**